karton_sqlmap/karton_sqlmap.py

```diff
@@ -102,37 +112,221 @@ def run(self, current_task: Task) -> None:
+                )
+            )
+
+        return results
+
+    @staticmethod
+    def _expand_urls_for_scanning(url: str) -> List[Tuple[str, str]]:
+        """
+        This converts a URL to a list of pairs:
+            - URL with injection point,
+            - original injection point value.
+
+        For example, 'https://example.com/path/file.html?id=1&q=2' would be converted to a list of:
+
+        [
+            ('https://example.com/path/file.html?id=1&q=*', '2'),
+            ('https://example.com/path/file.html?id=*&q=2', '1'),
+            ('https://example.com/path/*.html?id=1&q=2', 'file'),
+            ('https://example.com/*/file.html?id=1&q=2', 'path'),
+        ]
+        """
+        return sorted(
+            set(SQLmap._expand_query_parameters_for_scanning(url) + SQLmap._expand_path_segments_for_scanning(url))
+        )
+
+    def run(self, current_task: Task) -> None:
+        url = current_task.get_payload("url")
+        self.log.info("Requested to crawl and test SQL injection on %s", url)
+        url_parsed = urllib.parse.urlparse(url)
+
+        results = []
+
+        # Let's just try injecting example.com/[injection point]
+        root_injection_result = self._run_on_single_url(url + "*" if url.endswith("/") else url + "/*")
+        if root_injection_result:
+            results.append(root_injection_result)
+
+        response = http_requests.get(url)
+
+        # Unfortunately, crawling of clean URLs is not a feature that would get merged to sqlmap
+        # (https://github.com/sqlmapproject/sqlmap/issues/5561) so it is done here.
+        soup = BeautifulSoup(response.text)
+        urls_with_injection_points = set()
+        expanded_urls_with_example_values = []
+        for tag in soup.find_all():
+            new_url = None
+            for attribute in ["src", "href"]:
+                if attribute not in tag.attrs:
+                    continue
+
+                new_url = urllib.parse.urljoin(url, tag[attribute])
+                new_url_parsed = urllib.parse.urlparse(new_url)
+
+                if url_parsed.netloc == new_url_parsed.netloc:
+                    expanded_url = SQLmap._expand_urls_for_scanning(new_url)
+                    self.log.info("Found link %s, expanding to %s", new_url, expanded_url)
+
+                    for url_with_injection_point, example_value in expanded_url:
+                        if url_with_injection_point in urls_with_injection_points:
+                            continue
+
+                        urls_with_injection_points.add(url_with_injection_point)
+                        expanded_urls_with_example_values.append(
+                            (
+                                url_with_injection_point,
+                                example_value,
+                            )
+                        )
+
+        random.shuffle(expanded_urls_with_example_values)
+        expanded_urls_with_example_values = expanded_urls_with_example_values[
```

Comment:
I would like for loop to end if len(expanded_urls) > config_value, no need for it to loop through stuff taht would be truncated anyway

Reply:
but we perform random.shuffle() on that to get a random sample of the urls

without that, we wouldn't consider the urls that appear at the bottom of the page

and this loop is fast, it doesn't perform any server interaction